src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs

```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.Protocol.Plugins
+{
+    public static class PluginDiscoveryUtility
+    {
+        public static string InternalPluginDiscoveryRoot { get; set; } = null;
+
+        public static string GetInternalPlugins()
+        {
+            var rootDirectory = InternalPluginDiscoveryRoot ?? System.Reflection.Assembly.GetEntryAssembly()?.Location;;
+
+            return rootDirectory ?? Path.GetDirectoryName(rootDirectory);
+        }
+
+        public static string GetNuGetHomePluginsPath()
+        {
+            var nuGetHome = NuGetEnvironment.GetFolderPath(NuGetFolderPath.NuGetHome);
+
+            return Path.Combine(nuGetHome,
+                "plugins",
+#if IS_DESKTOP
+                "netfx"
+#else
+                "netcore"
+#endif
+                );
+        }
+
+        public static IEnumerable<string> GetConventionBasedPlugins(IEnumerable<string> directories)
+        {
+            var paths = new List<string>();
+            foreach (var directory in directories.Where(Directory.Exists))
```

Comment:
`foreach (var directory in directories.Where(Directory.Exists))`
This is really error prone design, since iterating a list and the same time if someone else updates this `directories` list then it will randomly throws `ArgumentException` without much details to debug it. I'd recommend first filtering `directories` list for exists list and then iterate.

Not directly related to this, but similar design issue - https://stackoverflow.com/questions/1167333/why-doesnt-a-foreach-loop-work-in-certain-cases